src/dir_contents.ml

```diff
@@ -43,6 +43,103 @@ end = struct
       , modules
       )
 
+  type field =
+    | Modules
+    | Intf_only
+
+  type incorrect_field =
+    { correct_field : field
+    ; module_: Module.t
+    }
+
+  type error =
+    | Incorrect_field of incorrect_field
+
+  let fold_errors ~f ~init ~modules ~intf_only =
+    let init =
+      Module.Name.Map.fold intf_only ~init
+        ~f:(fun (module_ : Module.t) acc ->
+          if Option.is_none module_.impl then
+            acc
+          else
+            f (Incorrect_field
+                 { correct_field = Modules
+                 ; module_
+                 }
+              ) acc)
+    in
+    Module.Name.Map.fold modules ~init
+      ~f:(fun (module_ : Module.t) acc ->
+        if Option.is_some module_.impl then
+          acc
+        else if not (Module.Name.Map.mem intf_only (Module.name module_)) then
+          f (Incorrect_field
+               { correct_field = Intf_only
+               ; module_
+               }
+            ) acc
+        else
+          acc)
+
+  let check_invalid_module_listing ~(buildable : Buildable.t) ~intf_only
+        ~modules ~modules_without_implementation_locs =
+    let (missing_modules, missing_intf_only) =
+      let (missing_modules, missing_intf_only) =
+        fold_errors ~init:([], []) ~modules ~intf_only
+          ~f:(fun e (missing_modules, missing_intf_only) ->
```

Comment:
Too bad you're not starting from a list, otherwise you could probably have simplified this using `filter_partition_map`.